libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp

```diff
@@ -0,0 +1,366 @@
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_CRSF, 0);
+    if (_uart) {
+        start_uart();
+    }
+#endif
+}
+
+AP_RCProtocol_CRSF::~AP_RCProtocol_CRSF() {
+    _singleton = nullptr;
+}
+
+void AP_RCProtocol_CRSF::process_pulse(uint32_t width_s0, uint32_t width_s1)
+{
+    uint8_t b;
+    if (ss.process_pulse(width_s0, width_s1, b)) {
+        _process_byte(ss.get_byte_timestamp_us(), b);
+    }
+}
+
+void AP_RCProtocol_CRSF::_process_byte(uint32_t timestamp_us, uint8_t byte)
+{
+    //debug("process_byte(0x%x)", byte);
+    // we took too long decoding, start again - the RX will only send complete frames so this is unlikely to fail
+    if (_frame_ofs > 0 && (timestamp_us - _start_frame_time_us) > CRSF_MAX_FRAME_TIME_US) {
+        _frame_ofs = 0;
+    }
+
+    _last_rx_time_us = timestamp_us;
+
+    // overflow check
+    if (_frame_ofs >= CRSF_FRAMELEN_MAX) {
+        _frame_ofs = 0;
+    }
+
+    // start of a new frame
+    if (_frame_ofs == 0) {
+        _start_frame_time_us = timestamp_us;
+    }
+
+    add_to_buffer(_frame_ofs++, byte);
+
+    // need a header to get the length
+    if (_frame_ofs < CSRF_HEADER_LEN) {
+        return;
+    }
+
+    // parse the length
+    if (_frame_ofs == CSRF_HEADER_LEN) {
+        // check for garbage frame
+        if (_frame.length > CRSF_FRAMELEN_MAX) {
+            _frame_ofs = 0;
+        }
+        return;
+    }
+
+    // overflow check
+    if (_frame_ofs > _frame.length + CSRF_HEADER_LEN) {
+        _frame_ofs = 0;
+        return;
+    }
+
+    // decode whatever we got and expect
+    if (_frame_ofs == _frame.length + CSRF_HEADER_LEN) {
+        log_data(AP_RCProtocol::CRSF, timestamp_us, (const uint8_t*)&_frame, _frame_ofs - CSRF_HEADER_LEN);
+
+        if ((timestamp_us - _last_frame_time_us) <= CRSF_INTER_FRAME_TIME_US_150HZ + CRSF_MAX_FRAME_TIME_US) {
+            _fast_telem = true;
+        } else {
+            _fast_telem = false;
+        }
+
+        // we consumed the partial frame, reset
+        _frame_ofs = 0;
+
+        uint8_t crc = crc8_dvb_s2(0, _frame.type);
+        for (uint8_t i = 0; i < _frame.length - 2; i++) {
+            crc = crc8_dvb_s2(crc, _frame.payload[i]);
+        }
+
+        // bad CRC
+        if (crc != _frame.payload[_frame.length - CSRF_HEADER_LEN]) {
+            return;
+        }
+
+        _last_frame_time_us = timestamp_us;
+        // decode here
+        if (decode_csrf_packet()) {
+            add_input(MAX_CHANNELS, _channels, false, -1);
+        }
+    }    
+}
+
+void AP_RCProtocol_CRSF::update(void)
+{
+    // if we are in standalone mode, process data from the uart
+    if (_uart) {
+        uint32_t now = AP_HAL::millis();
+        // for some reason it's necessary to keep trying to start the uart until we get data
+        if (now > _last_uart_start_time_ms + 1000U && _last_frame_time_us == 0) {
```

Comment:
this:
now > _last_uart_start_time_ms + 1000U
should be:
now - _last_uart_start_time_ms > 1000U
it makes a difference at the wrap point

Reply:
Fixed